dev/roadmap.md

```diff
@@ -1,103 +1,204 @@
-    - [ ] Joint Consensus - 安全地进行多个成员变更
-    - [ ] 多线程 Raftstore - 在多个线程处理不同 Region 的 Raft 逻辑
-    - [ ] 多线程 Apply Pool - 在多个线程执行不同 Region 已经提交了的命令
-+ Engine
-    - [ ] Titan - 把大的 key-values 从 LSM-Tree 中分离出来
-    - [ ] 可拔插的 Engine 接口 - 简化接口逻辑并且提供可扩展性
-+ Storage
-    - [ ] 在 scheduler 里做流控提前避免 write stall
-+ Transaction
-    - [x] 优化事务冲突
-    - [ ] 分布式 GC - 把 MVCC 垃圾回收的逻辑分布到 TiKV 控制
-+ Coprocessor
-    - [x] Streaming - 把大的数据集切成小块返回以减少内存消耗
-    - [ ] Chunk Execution - 按 chunk 的方式来处理数据以提高性能
-    - [ ] 请求跟踪 - 提供单个请求执行的详细信息
-+ Tools
-    - [x] TiKV Importer - 通过直接导入 SST 文件的方式加速数据导入
-+ Client
-    - [ ] 提供 Rust 版本的 TiKV client
-    - [ ] gRPC 消息批量化 - 减少消息交互的开销
-
-## PD
-
-- [x] Namespace 完善
-    - [x] 不同 Namespace 或者 Table 配置不同的副本策略
-- [x] Table Region 分散调度
-- [x] 调度支持优先级，更加可控
-- [ ] 使用机器学习优化调度
-- [ ] 优化 Region 元信息存储 - 把元信息存储在一个独立的存储引擎里
-
-## TiSpark
-
-- [ ] Limit/Order 下推
-- [x] DAG 接口接入（废除 Select 接口）
-- [ ] Index Join 和并行 merge join
-- [ ] Data Federation（桥接其他数据源，最好能和社区同步，这个接进来可以比较好扩展 Usecase，如果再做一个 InputFormat 适配就可以接 Hive 和 Presto 这些 Hadoop 上的数仓）
-
-## Tools
-
-- [x] 集群部署工具
-- [X] 高性能数据导入工具（lightning）
-- [X] 集群备份和恢复工具（包括全量+增量备份，mydumper + drainer/reparo）
-- [X] 改进 TiDB-Binlog 架构
-- [ ] 数据在线迁移工具（Syncer 升级版）
-- [ ] 集群诊断和分析工具
+### TiDB
+
+#### 功能
+
+* 支持 TiFlash 存储引擎
+* Optimizer Trace
+* Multi-Column Statistics
+* Normal CM-Sketch with TOPN Statistics
+* 完善 Plan Cache 功能
+* 简单的自适应 SQL 引擎
+* SQL Tuning Advisor
+* SQL Plan Management
+
+#### 性能
+
+* 提升 load CSV/data 性能
+* 提升 Prepare 语句性能
+* Generated Column 支持 Index
+* 优化 SQL 引擎部份算子
+    + 提升通过索引回表查询的性能
+    + Index Join 拆分为 Index Merge Join 和 Index Hash Join
+    + Radix Hash Join
+    + Index Merge
+    + Parallel Stream Aggregate
+    + Parallel Merge Sort
+    + Parallel Merge Join
+    + Full Vectorized Expression Evaluation
+* Indexes on Expressions
+* Multi-Index Scan
+* Join,Aggregate,Sort 算子支持外部存储
+* 优化执行引擎并发模型
+* 支持新 Cascades 优化器，支持 Cascades Planner，扩大优化器的搜索空间
+
+#### 易用性
+
+* 完善 Optimizer Hint 功能
+* 快速修复数据库或者表元信息及数据
+* 动态修改配置项
+* 空闲连接自动断开
+* 完善支持 MySQL 5.7 DDL 语句
+* 重构日志内容
+* 支持 admin checksum from … to … 校验数据完整性
+* 标准 SQL 查询 DDL 历史记录
+* 标准 SQL 管理 Binlog
+* 标准 SQL 管理集群
+* 多个 Ctrl 工具合并一个工具
+
+#### 高可用
+
+* Binlog 支持服务高可用
+
+### TiKV
+
+#### 功能
+
+* 集群规模最大支持 200+节点
+* 全量物理备份恢复
+* 动态 Split,Merge 热点 Region
+* 精细的内存控制
+* Raft
+    + Joint consensus
+    + 只读副本
+* 事务
+    + 悲观锁 GA
+    + 事务内不限制语句数量
+    + 支持 10GB 事务
+
+#### 性能
+
+* 提升 Scan 性能
+* 动态扩展工作线程数量
+* 弹性扩展只读副本
+* 优化调度系统确保 QPS 不抖动
+
+#### 易用性
+
+* 重构日志内容
+
+### TiFlash
+
+#### 功能
+
+* 列式存储
+* 通过 Raft learner 从 TiKV 同步数据
+* snapshot read
+
+### TiSpark
+
+#### 功能
+
+* 支持 batch write
+* 接入 TiFlash
+
+## Data Migration
+
+### 功能
+
+* 完善前向检查
+* 可视化管理同步规则
+* 可视化管理同步任务
+
+### 易用性
+
+* Ctrl 工具集成数据一致性校验功能
+* Ctrl 工具输出任务详细信息
+* 重构日志格式及内容
+
+### 高可用
+
+* 支持服务高可用
+
+## TiDB Toolkit
+
+### 功能
+
+* Loader 集成到 TiDB
+* TiDB Lightning 集成到 TiDB
+
+### 性能
+
+* TiDB Lightning 支持并行导入
+* TiDB Lightning 优化导入回放路径
+
+# TiDB 未来规划
```

Comment:
For the docs and docs-cn repos, we only support using one 1st-level heading.

Reply:
If it is not hierarchical, it will look very confusing

Reply:
OK. We can put Roadmap to the website footer and remove it from docs.